File: packages/amplify_datastore/android/src/main/kotlin/com/amazonaws/amplify/amplify_datastore/types/hub/FlutterOutboxMutationEnqueuedEvent.kt
// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

package com.amazonaws.amplify.amplify_datastore.types.hub

import com.amplifyframework.core.model.Model
import com.amplifyframework.datastore.syncengine.OutboxMutationEvent

class FlutterOutboxMutationEnqueuedEvent(
    override val eventName: String,
    private val modelName: String,
    private val element: OutboxMutationEvent.OutboxMutationEventElement<out Model>
) : FlutterHubEvent {
    override fun toValueMap(): Map<String, Any> {

        return mapOf(
            "eventName" to this.eventName,
            "modelName" to this.modelName,
            "element" to FlutterHubElement(this.element).toValueMap()
        )
    }
}

File: packages/amplify_datastore/ios/Classes/types/hub/FlutterOutboxMutationEnqueuedEvent.swift
// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import Foundation
import Amplify
import AmplifyPlugins

struct FlutterOutboxMutationEnqueuedEvent: FlutterHubEvent {
    var eventName: String
    var modelName: String
    var element: FlutterHubElement

    init(
        outboxMutationEnqueued: OutboxMutationEvent,
        eventName: String,
        modelSchemaRegistry: FlutterSchemaRegistry,
        customTypeSchemaRegistry: FlutterSchemaRegistry
    ) throws {
        self.eventName = shortEventName(eventName: eventName)
        self.modelName = outboxMutationEnqueued.modelName
        do {
            self.element = try FlutterHubElement(
                hubElement: outboxMutationEnqueued.element,
                modelSchemaRegistry: modelSchemaRegistry,
                customTypeSchemaRegistry: customTypeSchemaRegistry,
                modelName: modelName
            )
        } catch {
            throw FlutterDataStoreError.acquireSchemaForHub
        }
    }

    func toValueMap() -> [String: Any] {
        return [
            "eventName": eventName,
            "modelName": modelName,
            "element": element.toValueMap() as Any
        ]
    }
}
